File: benchmarks/agents/memory/PriorityTree.py
from decimal import Decimal
from random import uniform

import math
import torch



class PriorityTree:
    """
    A class storing the experience priorities.
    """

    def __init__(self, capacity, initial_priority, n_children):
        """
        Create a priority tree.
        :param capacity: the tree's capacity
        :param initial_priority: the initial priority given to first elements
        :param n_children: the number of children each node has
        """

        # Store the priority tree parameters.
        self.initial_priority = initial_priority
        self.capacity = capacity
        self.n_children = n_children

        # Robust computation of the trees' depth.
        self.depth = math.floor(math.log(self.capacity, n_children))
        if int(math.pow(n_children, self.depth)) < self.capacity:
            self.depth += 1

        # Create a tensor of priorities, an empty sum-tree and an empty max-tree.
        self.priorities = torch.zeros([self.capacity])
        self.sum_tree = self.create_sum_tree(self.depth, n_children)
        self.max_tree = self.create_max_tree(self.depth, n_children)
        self.current_id = 0

    @staticmethod
    def create_sum_tree(depth, n_children):
        """
        Create a sum-tree.
        Importantly, tree elements must be python floats to avoid numerical precision error of pytorch tensors.
        :param depth: the tree's depth
        :param n_children: the number of children each node has
        :return: the tree
        """
        return [
            [0.0 for _ in range(int(math.pow(n_children, i)))]
            for i in reversed(range(depth))
        ]

    @staticmethod
    def create_max_tree(depth, n_children):
        """
        Create a max-tree.
        :param depth: the tree's depth
        :param n_children: the number of children each node has
        :return: the tree
        """
        return [
            torch.zeros([int(math.pow(n_children, i))])
            for i in reversed(range(depth))
        ]

    def sum(self):
        """
        Compute the sum of all priorities.
        :return: the sum of all priorities
        """
        if self.current_id == 0:
            return 0
        return self.sum_tree[-1][0]

    def max(self):
        """
        Find the largest priority.
        :return: the largest priority
        """
        if self.current_id == 0:
            return self.initial_priority
        return self.max_tree[-1][0].item()

    def clear(self):
        """
        Empty the priority tree.
        """
        self.current_id = 0
        self.priorities = torch.zeros([self.capacity])
        self.sum_tree = self.create_sum_tree(self.depth, self.n_children)
        self.max_tree = self.create_max_tree(self.depth, self.n_children)

    def __len__(self):
        """
        Retrieve the number of experiences stored in the buffer.
        :return: the number of experiences stored in the buffer
        """
        return min(self.current_id, self.capacity)

    def append(self, priority):
        """
        Add a priority in the priority tree.
        :param priority: the new priority
        """
        index = self.current_id % self.capacity
        self.update_sum_tree(index, priority)
        self.update_max_tree(index, priority)
        self.priorities[index] = priority
        self.current_id += 1

    def __getitem__(self, index):
        """
        Retrieve an element from the list.
        :param index: the index of the experience whose priority must be retrieved
        :return: the element
        """
        return self.priorities[self.internal_index(index)]

    def __setitem__(self, index, priority):
        """
        Replace an element in the list.
        :param index: the index of the experience whose priority must be replaced
        :param priority: the new priority
        """
        index = self.internal_index(index)
        self.update_sum_tree(index, priority)
        self.update_max_tree(index, priority)
        self.priorities[index] = priority

    def internal_index(self, index):
        """
        Transform an experience index to its internal index
        :param index: the experience index
        :return: the internal index
        """
        if self.current_id >= self.capacity:
           index += self.current_id
        return index % self.capacity

    def external_index(self, index):
        """
        Transform an internal index to its experience index
        :param index: the internal index
        :return: the experience index
        """
        if self.current_id >= self.capacity:
            index -= (self.current_id % self.capacity)
        return index % self.capacity

    def sample_indices(self, n):
        """
        Sample indices of buffer elements proportionally to their priorities.
        :param n: the number of indices to sample
        :return: the sampled indices
        """

        # Sample 'n' indices with a probability proportional to their priorities.
        indices = []
        for i in range(n):

            # Sample a priority between zero and the sum of priorities.
            sampled_priority = uniform(0, self.sum())

            # Compute the element index associated to the sampled priority using inverse transform sampling.
            indices.append(self.tower_sampling(sampled_priority))
        return indices

    def tower_sampling(self, priority):
        """
        Compute the experience index associated to the sampled priority using inverse transform sampling (tower sampling).
        :param priority: the sampled priority
        :return: the experience index
        """

        # If the priority is larger than the sum of priorities, return the index of the last element.
        if priority > self.sum():
            return self.external_index(len(self) - 1)

        # Go down the sum-tree until the leaf node is reached.
        index = 0
        for level in reversed(range(-1, self.depth - 1)):

            # Iterate over the children of the current node, keeping track of the sum of priorities.
            total = 0
            for i in range(self.n_children):

                # Get the priority of the next child.
                child_index = self.n_children * index + i
                new_priority = self.priorities[child_index] if level == -1 else self.sum_tree[level][child_index]

                # If the priority is about to be superior to the total, stop iterating over the children.
                if priority <= total + new_priority:
                    index = child_index
                    priority -= total
                    break

                # Otherwise, increase the sum of priorities.
                total += new_priority

        # Return the element index corresponding to the sampled priority.
        return self.external_index(index)

    def parent_index(self, idx):
        """
        Compute the index of the parent element.
        :param idx: the index of the element whose parent index must be computed
        :return: the parent index
        """
        return idx if idx < 0 else idx // self.n_children

    def update_sum_tree(self, index, new_priority):
        """
        Update the sum-tree to reflect an element being set to a new priority.
        :param index: the internal index of the element
        :param new_priority: the new priority
        """

        # Compute the parent index.
        parent_index = self.parent_index(index)

        # Go up the tree until the root node is reached.
        depth = 0
        priority = self.priorities[index].item()
        while depth < self.depth:

            # Update the sums in the sum-tree.
            diff = Decimal(self.sum_tree[depth][parent_index]) + Decimal(new_priority) - Decimal(priority)
            self.sum_tree[depth][parent_index] = float(diff)

            # Update parent indices and tree depth.
            depth += 1
            parent_index = self.parent_index(parent_index)

    def update_max_tree(self, index, new_priority):
        """
        Update the max-tree to reflect an element being set to a new priority.
        :param index: the internal index of the element
        :param new_priority: the new priority
        """

        # Compute the parent index and the old priority.
        parent_index = self.parent_index(index)
        old_priority = self.priorities[index].item()

        # Go up the tree until the root node is reached.
        depth = 0
        while depth < self.depth:

            # Update the maximum values in the max-tree.
            parent_value = self.max_tree[depth][parent_index].item()
            if parent_value == old_priority:
                self.max_tree[depth][parent_index] = self.max_child_value(depth, parent_index, index, old_priority, new_priority)
            elif parent_value < new_priority:
                self.max_tree[depth][parent_index] = new_priority
            else:
                break

            # Update parent indices and tree depth.
            depth += 1
            parent_index = self.parent_index(parent_index)

    def max_child_value(self, depth, parent_index, index, old_priority, new_priority):
        """
        Compute the maximum value among the child nodes.
        :param depth: the depth of the parent node
        :param parent_index: the internal index of the parent node
        :param index: the internal index of the node whose value is being set to a new priority
        :param old_priority: the old priority
        :param new_priority: the new priority
        :return: the maximum value
        """
        first_child = self.n_children * parent_index
        if depth == 0:
            children = self.priorities[first_child: first_child + self.n_children]
            children[index - first_child] = new_priority
            max_value = children.max()
            children[index - first_child] = old_priority
        else:
            max_value = self.max_tree[depth - 1][first_child: first_child + self.n_children].max()
        return max_value

    @staticmethod
    def to_str(tree):
        """
        Create a string representation of the tree passed as parameters.
        :param tree: the tree
        :return: a string representing a tree
        """
        tree_str = []
        for row in tree:
            tree_str.append(row.tolist())
        return f"{tree_str}"

File: benchmarks_cpp/src/priority_tree.cpp
#include "priority_tree.hpp"
#include <cmath>

using namespace torch::indexing;

PriorityTree::PriorityTree(int capacity, float initial_priority, int n_children) {

    // Store the priority tree parameters.
    this->initial_priority = initial_priority;
    this->capacity = capacity;
    this->n_children = n_children;

    // Robust computation of the trees' depth.
    this->depth = std::floor(std::log(this->capacity) / std::log(n_children));
    if (static_cast<int>(std::pow(n_children, this->depth)) < this->capacity) {
        this->depth += 1;
    }

    // Create a tensor of priorities, an empty sum-tree and an empty max-tree.
    this->priorities = torch::zeros({this->capacity});
    this->sum_tree = this->createSumTree(this->depth, n_children);
    this->max_tree = this->createMaxTree(this->depth, n_children);
    this->current_id = 0;
    this->need_refresh_all = true;
}

SumTree PriorityTree::createSumTree(int depth, int n_children) {
    SumTree tree;

    for (auto i = depth - 1; i >= 0; i--) {
        int n = std::pow(n_children, i);
        std::vector<long double> row(n);
        tree.push_back(std::move(row));
    }
    return tree;
}

MaxTree PriorityTree::createMaxTree(int depth, int n_children) {
    MaxTree tree;

    for (auto i = depth - 1; i >= 0; i--) {
        tree.push_back(torch::zeros({static_cast<int>(std::pow(n_children, i))}));
    }
    return tree;
}

long double PriorityTree::sum() {
    if (this->current_id == 0) {
        return 0;
    }
    return this->sum_tree[this->sum_tree.size() - 1][0];
}

float PriorityTree::max() {
    if (this->current_id == 0) {
        return this->initial_priority;
    }
    return this->max_tree[this->max_tree.size() - 1][0].item<float>();
}

void PriorityTree::clear() {
    this->current_id = 0;
    this->need_refresh_all = true;
    this->priorities = torch::zeros({this->capacity});
    this->sum_tree = this->createSumTree(this->depth, this->n_children);
    this->max_tree = this->createMaxTree(this->depth, this->n_children);
}

int PriorityTree::size() {
    return std::min(this->current_id, this->capacity);
}

void PriorityTree::append(float priority) {

    int idx = this->current_id % this->capacity;
    float old_priority = this->priorities[idx].item<float>();

    // Add a new priority to the list of priorities.
    this->priorities[idx] = priority;
    this->current_id += 1;
    this->updateMaxTree(idx, old_priority);
    this->updateSumTree(idx, old_priority);

    // Check if the full sum tree must be refreshed.
    if (this->max() != this->initial_priority and this->need_refresh_all == true) {
        this->refreshAllSumTree();
        this->need_refresh_all = false;
    }
}

float PriorityTree::get(int index) {
    return this->priorities[this->internalIndex(index)].item<float>();
}

void PriorityTree::set(int index, float priority) {

    int idx = this->internalIndex(index);
    float old_priority = this->priorities[idx].item<float>();

    // Replace the old priority with the new priority.
    this->priorities[idx] = priority;
    this->updateMaxTree(idx, old_priority);
    this->updateSumTree(idx, old_priority);

    // Check if the full sum tree must be refreshed.
    if (this->max() != this->initial_priority and this->need_refresh_all == true) {
        this->refreshAllSumTree();
        this->need_refresh_all = false;
    }
}

int PriorityTree::internalIndex(int index) {
    if (this->current_id >= this->capacity) {
       index += this->current_id;
    }
    index %= this->capacity;
    return (index >= 0) ? index : index + this->size();
}

int PriorityTree::externalIndex(int index) {
    if (this->current_id >= this->capacity) {
        index -= (this->current_id % this->capacity);
    }
    index %= this->capacity;
    return (index >= 0) ? index : index + this->size();
}

torch::Tensor PriorityTree::sampleIndices(int n) {

    // Sample priorities between zero and the sum of priorities.
    torch::Tensor sampled_priorities = torch::rand({n}) * static_cast<float>(this->sum());

    // Sample 'n' indices with a probability proportional to their priorities.
    torch::Tensor indices = torch::zeros({n}, torch::kInt64);
    for (auto i = 0; i < n; i++) {
        float priority = sampled_priorities.index({i}).item<float>();
        indices.index_put_({i}, static_cast<long>(this->towerSampling(priority)));
    }
    return indices;
}

int PriorityTree::towerSampling(float priority) {

    // If the priority is larger than the sum of priorities, return the index of the last element.
    if (priority > this->sum()) {
        return this->externalIndex(this->size() - 1);
    }

    // Go down the sum-tree until the leaf node is reached.
    float new_priority = 0;
    int index = 0;
    for (int level = this->depth - 2; level >= -1; level--) {

        // Iterate over the children of the current node, keeping track of the sum of priorities.
        float total = 0;
        for (auto i = 0; i < this->n_children; i++) {

            // Get the priority of the next child.
            int child_index = this->n_children * index + i;
            if (level == -1) {
                new_priority = this->priorities[child_index].item<float>();
            } else {
                new_priority = static_cast<float>(this->sum_tree[level][child_index]);
            }

            // If the priority is about to be superior to the total, stop iterating over the children.
            if (priority <= total + new_priority) {
                index = child_index;
                priority -= total;
                break;
            }

            // Otherwise, increase the sum of priorities.
            total += new_priority;
        }
    }

    // Return the element index corresponding to the sampled priority.
    return this->externalIndex(index);
}

int PriorityTree::parentIndex(int idx) {
    return (idx < 0) ? idx : idx / this->n_children;
}

void PriorityTree::updateSumTree(int index, float old_priority) {

    // Compute the parent index.
    int parent_index = this->parentIndex(index);

    // Go up the tree until the root node is reached.
    int depth = 0;
    float new_priority = this->priorities[index].item<float>();
    while (depth < this->depth) {

        // Update the sums in the sum-tree.
        this->sum_tree[depth][parent_index] += new_priority - old_priority;

        // Update parent indices and tree depth.
        depth += 1;
        parent_index = this->parentIndex(parent_index);
    }
}

void PriorityTree::refreshAllSumTree() {

    // Fill the sum-tree with zeros.
    this->sum_tree = this->createSumTree(this->depth, this->n_children);

    // Iterate over all the priorities.
    for (auto index = 0; index < this->size(); index++) {

        // Compute the parent index and current priority.
        int parent_index = this->parentIndex(index);
        float priority = this->priorities[index].item<float>();

        // Go up the tree until the root node is reached.
        int depth = 0;
        while (depth < this->depth) {

            // Update the sums in the sum-tree.
            this->sum_tree[depth][parent_index] += priority;

            // Update parent indices and tree depth.
            depth += 1;
            parent_index = this->parentIndex(parent_index);
        }
    }
}

void PriorityTree::updateMaxTree(int index, float old_priority) {

    // Compute the parent index and the old priority.
    int parent_index = this->parentIndex(index);
    float new_priority = this->priorities[index].item<float>();

    // Go up the tree until the root node is reached.
    int depth = 0;
    while (depth < this->depth) {

        // Update the maximum values in the max-tree.
        float parent_value = this->max_tree[depth][parent_index].item<float>();
        if (parent_value == old_priority) {
            this->max_tree[depth][parent_index] = this->maxChildValue(depth, parent_index, index, old_priority, new_priority);
        } else if (parent_value < new_priority) {
            this->max_tree[depth][parent_index] = new_priority;
        } else {
            break;
        }

        // Update parent indices and tree depth.
        depth += 1;
        parent_index = this->parentIndex(parent_index);
    }
}

float PriorityTree::maxChildValue(int depth, int parent_index, int index, float old_priority, float new_priority) {
    int first_child = this->n_children * parent_index;
    auto slice = Slice(first_child, first_child + this->n_children);
    float max_value = 0;

    if (depth == 0) {
        torch::Tensor children = this->priorities.index({slice});
        max_value = children.max().item<float>();
    } else {
        max_value = this->max_tree[depth - 1].index({slice}).max().item<float>();
    }
    return max_value;
}

std::string PriorityTree::maxTreeToStr() {

    int n = static_cast<int>(this->max_tree.size());
    std::string tree_str = "[";

    // Iterate over all sub-lists.
    for (auto i = 0; i < n; i++) {

        // Open the bracket in the string.
        if (i != 0)
            tree_str += ", [";
        else
            tree_str += "[";

        // Iterate over all elements.
        int m = this->max_tree[i].numel();
        for (auto j = 0; j < m; j++) {

            // Add all elements to the string.
            if (j != 0)
                tree_str += ", ";
            tree_str += this->toString(this->max_tree[i].index({j}).item<float>());
        }

        // Close the bracket in the string.
        tree_str += "]";
    }
    return tree_str + "]";
}

std::string PriorityTree::sumTreeToStr() {

    int n = static_cast<int>(this->sum_tree.size());
    std::string tree_str = "[";

    // Iterate over all sub-lists.
    for (auto i = 0; i < n; i++) {

        // Open the bracket in the string.
        if (i != 0)
            tree_str += ", [";
        else
            tree_str += "[";

        // Iterate over all elements.
        int m = static_cast<int>(this->sum_tree[i].size());
        for (auto j = 0; j < m; j++) {

            // Add all elements to the string.
            if (j != 0)
                tree_str += ", ";
            tree_str += this->toString(static_cast<float>(this->sum_tree[i][j]));
        }

        // Close the bracket in the string.
        tree_str += "]";
    }
    return tree_str + "]";
}

std::string PriorityTree::toString(float value, int precision) {
    std::ostringstream out;
    out.precision(precision);
    out << std::fixed << value;
    return out.str();
}
